src/main/scala/cognite/spark/v1/AssetHierarchyBuilder.scala

```diff
@@ -270,16 +189,28 @@ class AssetHierarchyBuilder(config: RelationConfig)(val sqlContext: SQLContext)
     }
   }
 
-  def delete(toDelete: Vector[Asset], deleteMissingAssets: Boolean): IO[Unit] =
+  def deleteMissingChildren(trees: Vector[AssetSubtree], deleteMissingAssets: Boolean): IO[Unit] =
     if (deleteMissingAssets) {
-      batchedOperation[Long, Nothing](
-        toDelete.map(_.id),
-        idBatch =>
-          client.assets
-            .deleteByIds(idBatch, recursive = true, ignoreUnknownIds = true)
-            .flatTap(_ => incMetrics(itemsDeleted, idBatch.length))
-            .as(Vector())
-      ).void
+      val ingestedNodeSet = trees.flatMap(_.allNodes).map(_.externalId).toSet
+      // list all subtrees of the tree root and filter those which are not in the ingested set
+      for {
+        idsToDelete <- batchedOperation[String, Long](
+          trees.map(_.root.externalId),
+          ids =>
+            client.assets.filter(AssetsFilter(assetSubtreeIds = Some(ids.map(CogniteExternalId))))
+              .compile.toVector
+              .map(_.filter(a => !a.externalId.exists(ingestedNodeSet.contains))
+                .map(_.id))
+        )
```

Comment:
I only fetch the subtrees in the delete mode (which is not used in Jetfire). This means we'll fetch the nodes twice, but we can easily batch multiple subtrees together. For the other uses, we'd have to reaggregate them together ...